compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -162,10 +162,17 @@ fn fmt_printer<'a, 'tcx>(infcx: &'a InferCtxt<'tcx>, ns: Namespace) -> FmtPrinte
         let mut infcx_inner = infcx.inner.borrow_mut();
         let ty_vars = infcx_inner.type_variables();
         let var_origin = ty_vars.var_origin(ty_vid);
-        if let TypeVariableOriginKind::TypeParameterDefinition(name, _) = var_origin.kind
+        if let TypeVariableOriginKind::TypeParameterDefinition(name, def_id) = var_origin.kind
             && !var_origin.span.from_expansion()
         {
-            Some(name)
+            let ty::Generics { params, .. } = infcx.tcx.generics_of(infcx.tcx.parent(def_id));
+            if let Some(generic_param_def) = params.iter().find(|&generic_param_def| generic_param_def.name == name)
+                && let ty::GenericParamDefKind::Type { synthetic: true, .. } = generic_param_def.kind
```

Comment:
please don't look up by name, instead check that the `DefId` is equal. Instead use `param_def_id_to_index` and then `param_at` of the `generics`.